### PR TITLE
Refactored input controls

### PR DESCRIPTION
**Pull request by sorinsarca**
## Description

Contains the changes from #3013 (only for checkbox, switch, radio and select).

### CODE REVIEW

File: src/util/helpers.js [hunk 1]
```diff
@@ -99,21 +99,63 @@ export function addOnceEventListener (el, event, cb) {
...
+  if (last < 0) return obj === undefined ? fallback : obj
+
+  for (let i = 0; i < last; i++) {
+    if (obj === null || obj === undefined) {
+      return fallback
+    }
+    obj = obj[path[i]]
+  }
+
+  if (obj === null || obj === undefined) return fallback
```
Comment (KaelWD) [inline src/util/helpers.js]: `== null` is the same as `=== null || === undefined`

File: src/util/helpers.js [hunk 2]
```diff
@@ -99,21 +99,63 @@ export function addOnceEventListener (el, event, cb) {
...
 }
 
-export function getObjectValueByPath (obj, path) {
+export function getNestedValue (obj, path, fallback) {
+  const last = path.length - 1
+
+  if (last < 0) return obj === undefined ? fallback : obj
+
+  for (let i = 0; i < last; i++) {
+    if (obj === null || obj === undefined) {
```
Comment (KaelWD) [inline src/util/helpers.js]: Again here

File: src/util/helpers.js [hunk 3]
```diff
@@ -99,21 +99,63 @@ export function addOnceEventListener (el, event, cb) {
...
+    return false
+  }
+
+  return props.every(p => looseEqual(a[p], b[p]))
+}
+
+export function getObjectValueByPath (obj, path, fallback) {
   // credit: http://stackoverflow.com/questions/6491463/accessing-nested-javascript-objects-with-string-key#comment55278413_6491621
-  if (!path || path.constructor !== String) return
+  if (!path || path.constructor !== String) return fallback
```
Comment (KaelWD) [inline src/util/helpers.js]: Any reason you're using `constructor` rather than `typeof`?

File: src/util/helpers.js [hunk 4]
```diff
@@ -99,21 +99,63 @@ export function addOnceEventListener (el, event, cb) {
...
-    } else {
-      return
-    }
-  }
-  return obj
+  return getNestedValue(obj, path.split('.'), fallback)
+}
+
+export function getPropertyFromItem (item, property, fallback) {
+  if (property === null || property === undefined) return item === undefined ? fallback : item
```
Comment (KaelWD) [inline src/util/helpers.js]: Split into two ifs, without the ternary

File: src/util/helpers.js [hunk 5]
```diff
@@ -99,21 +99,63 @@ export function addOnceEventListener (el, event, cb) {
...
-      obj = obj[k]
-    } else {
-      return
-    }
-  }
-  return obj
+  return getNestedValue(obj, path.split('.'), fallback)
+}
+
+export function getPropertyFromItem (item, property, fallback) {
```
Comment (KaelWD) [inline src/util/helpers.js]: Maybe add a comment explaining what this is for. Why does it take a function? The function signature should probably be added to the type definition too.

File: src/util/helpers.js [hunk 6]
```diff
@@ -99,21 +99,63 @@ export function addOnceEventListener (el, event, cb) {
...
+
+  if (property.constructor === String) return getObjectValueByPath(item, property, fallback)
+
+  if (Array.isArray(property)) return getNestedValue(item, property, fallback)
+
+  if (typeof property !== 'function') return fallback
+
+  const value = property(item, fallback)
+
+  return typeof value === "undefined" ? fallback : value
```
Comment (KaelWD) [inline src/util/helpers.js]: Use single quotes.

File: src/util/helpers.js (hunk 3 above)
Comment (sorinsarca) [reply src/util/helpers.js]: class X extends String {}
Anyway, it was like that before my commits. So I really don't know the reason.

File: src/util/helpers.js [hunk 7]
```diff
@@ -99,21 +99,63 @@ export function addOnceEventListener (el, event, cb) {
...
+    }
+    obj = obj[path[i]]
+  }
+
+  if (obj === null || obj === undefined) return fallback
+
+  return obj[path[last]] === undefined ? fallback : obj[path[last]]
+}
+
+export function looseEqual (a, b) {
```
[the remark this answers is not in the document]
Comment (sorinsarca) [reply src/util/helpers.js]: Can be called like that

File: src/util/helpers.js (hunk 3 above)
Comment (KaelWD) [reply src/util/helpers.js]: Ah right, didn't notice that in the diff.

File: src/mixins/selectable.js [hunk 8]
```diff
@@ -29,12 +30,16 @@ export default {
     id: String,
     inputValue: null,
     falseValue: null,
+    trueValue: null,
+    multiple: {
+      type: Boolean,
+      default: null
```
[the remark this answers is not in the document]
Comment (sorinsarca) [reply src/mixins/selectable.js]: Yes, setting multiple to false and using array as a single value

File: src/util/helpers.js (hunk 5 above)
Comment (sorinsarca) [reply src/util/helpers.js]: Sometimes only a path is not enough to get the desired property. The function receives as arg the item.

File: src/components/VSelect/VSelectList.js [hunk 9]
```diff
@@ -144,23 +144,16 @@ export default {
...
-      return getObjectValueByPath(item, this.itemAvatar)
+      return Boolean(getPropertyFromItem(item, this.itemAvatar, false))
     },
     getDisabled (item) {
-      return getObjectValueByPath(item, this.itemDisabled)
+      return Boolean(getPropertyFromItem(item, this.itemDisabled, false))
     },
     getText (item) {
-      return this.getPropertyFromItem(item, this.itemText)
+      return (getPropertyFromItem(item, this.itemText, item) || '').toString();
```
Comment (KaelWD) [inline src/components/VSelect/VSelectList.js]: Maybe `String()...`? depends how we want to handle falsy values I guess. 

Remove the semicolon too please.

File: src/util/helpers.js [hunk 10]
```diff
@@ -99,21 +99,63 @@ export function addOnceEventListener (el, event, cb) {
...
-    } else {
-      return
-    }
-  }
-  return obj
+  return getNestedValue(obj, path.split('.'), fallback)
+}
+
+export function getPropertyFromItem (item, property, fallback) {
+  if (property == null) return item === undefined ? fallback : item
```
Comment (KaelWD) [inline src/util/helpers.js]: I meant 
```js
if (property === undefined) return fallback
if (property === null) return item
```

**Comment by johnleider** [comment]
Merging this in as I need to progress with aspects that this touches and I don't want to keep breaking it.

**Pull request merged**